### PR TITLE
fix: Refactor MassScannersCron to handle flexible IP formats. Closes #678

### LINKED ISSUE #678: Thousands of warnings from MassScannersCron
On every run of the  `MassScannersCron` job, I am getting a lot of warning messages. On the last one there were 4254 in the `greedybear.log` file: 

```
2026-01-04 04:03:00,021 - greedybear.cronjobs.base.MassScannersCron - execute - INFO - Starting execution
2026-01-04 04:03:00,204 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line:     
2026-01-04 04:03:00,212 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line:     
2026-01-04 04:03:00,629 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line:     
2026-01-04 04:03:13,806 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::1008:2001 # binaryedge.ninja
2026-01-04 04:03:13,806 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::100a:1 # binaryedge.ninja
2026-01-04 04:03:13,809 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::1016:f001 # binaryedge.ninja
2026-01-04 04:03:13,809 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::1020:6001 # binaryedge.ninja
2026-01-04 04:03:13,809 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::1034:b001 # binaryedge.ninja
2026-01-04 04:03:13,809 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::103:2001 # binaryedge.ninja
2026-01-04 04:03:13,810 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::104d:6001 # binaryedge.ninja
2026-01-04 04:03:13,810 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::104e:2001 # binaryedge.ninja
2026-01-04 04:03:13,810 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::105f:1 # binaryedge.ninja
2026-01-04 04:03:13,810 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::106e:d001 # binaryedge.ninja
2026-01-04 04:03:13,810 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::107b:1001 # binaryedge.ninja
2026-01-04 04:03:13,810 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::1087:7001 # binaryedge.ninja
2026-01-04 04:03:13,810 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::10a7:f001 # binaryedge.ninja
2026-01-04 04:03:13,811 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::10aa:1001 # binaryedge.ninja
2026-01-04 04:03:13,811 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::10ae:d001 # binaryedge.ninja
2026-01-04 04:03:13,811 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::10c0:b001 # binaryedge.ninja
2026-01-04 04:03:13,811 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::10d5:a001 # binaryedge.ninja
2026-01-04 04:03:13,811 - greedybear.cronjobs.base.MassScannersCron - run - WARNING - unexpected line: 2400:6180:0:d0::10dc:f001 # binaryedge.ninja
...
```

Does anybody else see this on their instance?

### PR DESCRIPTION
**Pull request by opbot-xd**
# Description

This PR refactors the [MassScannersCron](cci:2://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/mass_scanners.py:9:0-61:22) job to handle diverse IP address formats from the external data source, eliminating false "unexpected line" warnings.

**Problem:** The old strict regex pattern (`IPv4 # reason`) flagged valid data as "unexpected" (plain IPs, IPv6, special strings).

**Solution:** Implemented both suggestions from issue discussion:
1. **Flexible IP validation** - Simple regex extracts candidates, [is_valid_ipv4()](cci:1://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/utils.py:133:0-148:26) validates them
2. **Reduced log level** - Changed WARNING → DEBUG for non-IP lines (expected from external source)

**Changes:**
- Added [is_valid_ipv4()](cci:1://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/utils.py:133:0-148:26) utility in [greedybear/cronjobs/extraction/utils.py](cci:7://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/utils.py:0:0-0:0) (DRY principle)
- Refactored [get_ioc_type()](cci:1://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/utils.py:151:0-162:37) to use new utility
- Updated [MassScannersCron](cci:2://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/mass_scanners.py:9:0-61:22) parser with flexible extraction + validation
- Added 26 comprehensive tests (11 for [is_valid_ipv4()](cci:1://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/utils.py:133:0-148:26), 15 for [MassScannersCron](cci:2://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/mass_scanners.py:9:0-61:22))

Now correctly handles:
- Plain IPs: `45.83.67.252` ✅
- IPs with comments: `192.168.1.100 # normal comment` ✅  
- IPv6 (skipped at DEBUG): `2001:db8::1` ✅
- Invalid strings (skipped at DEBUG): `/w00tw00t.at.ISC.SANS.DFind:)` ✅

## Related issues
Closes #678

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowlproject.github.io/docs/GreedyBear/Contribute/) to this project.
- [x] The pull request is for the branch `develop`.
- [ ] I have added documentation of the new features.
- [x] Linter (`Ruff`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/GreedyBear/Contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved. All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/docs/blob/main/docs/GreedyBear/Contribute.md)).

### CODE REVIEW

**Comment by opbot-xd** [comment]
## Test Coverage Rationale

Added 26 comprehensive tests because:

1. **External data variability** - The upstream source contains diverse formats (plain IPs, IPs with comments, IPv6, random strings) that we don't control
2. **Core utility function** - [is_valid_ipv4()](cci:1://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/utils.py:133:0-148:26) is now used by multiple modules, requiring thorough edge case validation
3. **Regression prevention** - Ensures the flexible parser correctly accepts valid IPs while rejecting invalid inputs
4. **Documentation** - Tests serve as living documentation of expected input formats and behavior

All 67 tests pass, giving confidence this handles real-world threat intelligence data correctly. ✅

**Comment by opbot-xd** [comment]
@regulartim Please review this PR when you get time...

File: greedybear/cronjobs/mass_scanners.py [hunk 1]
```diff
@@ -3,12 +3,17 @@
...
 from greedybear.models import IOC, MassScanner
 
 
 class MassScannersCron(Cronjob):
     def run(self) -> None:
-        regex_compiled = re.compile(r"(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})\s*#\s*(.+)*", re.DOTALL)
+        # Simple regex to extract potential IPv4 addresses
+        ip_candidate_regex = re.compile(r"(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})")
+        # Regex to extract optional comment/reason after '#'
+        comment_regex = re.compile(r"#\s*(.+)", re.DOTALL)
```
Comment (regulartim) [inline greedybear/cronjobs/mass_scanners.py]: I know that this was in the code already, but why do we have re.DOTALL here? To capture comments spanning multiple lines? @mlodic

File: greedybear/cronjobs/mass_scanners.py [hunk 2]
```diff
@@ -18,17 +23,34 @@ def run(self) -> None:
...
+                comment_match = comment_regex.search(line)
+                if comment_match:
+                    reason = comment_match.group(1)
+
+                # Add or update mass scanner entry
+                try:
+                    MassScanner.objects.get(ip_address=ip_address)
+                except MassScanner.DoesNotExist:
+                    self.log.info(f"added new mass scanner {ip_address}")
+                    MassScanner(ip_address=ip_address, reason=reason).save()
```
Comment (regulartim) [inline greedybear/cronjobs/mass_scanners.py:52]: Maybe switch the position of those two lines. I think it's better to log after the `save()` call.

**Review by regulartim — CHANGES_REQUESTED** [review]
Sorry for being late to the party. Very nice work @opbot-xd ! Just a few minor things.

File: greedybear/cronjobs/mass_scanners.py (hunk 1 above)
Comment (opbot-xd) [reply greedybear/cronjobs/mass_scanners.py]: Good catch! You're right - `re.DOTALL` is likely unnecessary here since we're already processing the data line-by-line with `r.iter_lines()`. 

The comment extraction happens on individual lines, so newlines within a single line shouldn't occur. I kept `re.DOTALL` from the original implementation to maintain the same behavior, but it's probably redundant.

I can remove it and simplify to:
```python
comment_regex = re.compile(r"#\s*(.+)")

File: greedybear/cronjobs/mass_scanners.py (hunk 2 above)
Comment (opbot-xd) [reply greedybear/cronjobs/mass_scanners.py:52]: Makes sense! Logging should happen after successful save in case the database operation fails. I'll swap them:
```python
MassScanner(ip_address=ip_address, reason=reason).save()
self.log.info(f"added new mass scanner {ip_address}")

File: greedybear/cronjobs/mass_scanners.py (hunk 1 above)
Comment (regulartim) [reply greedybear/cronjobs/mass_scanners.py]: Let's see what @mlodic says. He surely had a reason to put it in.

Comment (opbot-xd) [reply greedybear/cronjobs/mass_scanners.py]: I investigated whether multi-line comments actually occur:
**Finding:** They don't. After checking the [actual data source](https://raw.githubusercontent.com/stamparm/maltrail/master/trails/static/mass_scanner.txt), the format is strictly:
IP # single-line comment

**Why `re.DOTALL` can't help anyway:**
- `r.iter_lines()` splits the HTTP response by newlines
- Each iteration processes one line (newlines are delimiters, not content)
- Even if a comment had `\n` in it, we'd never see it - it would be split into separate lines before reaching our regex

Comment (opbot-xd) [reply greedybear/cronjobs/mass_scanners.py]: Sorry for the confusion! When I made the changes, I only saw the two specific review comments (about `re.DOTALL` and log placement), so I addressed both and pushed. 

Your comment asking to defer to @mlodic loaded after I had already committed and pushed - GitHub timing issue on my end. (forgot to refresh the page.) 😅

I've already pushed the changes addressing both review points, but I'm happy to revert if @mlodic prefers to keep `re.DOTALL` for any reason I'm not aware of. Just let me know!

Comment (mlodic) [reply greedybear/cronjobs/mass_scanners.py]: thanks for letting me check but honestly I don't remember. Let's keep your change
